### ISSUE TITLE
IAsyncEnumerableOfTConverter<TAsyncEnumerable, TElement> throws OutOfMemoryException when custom JsonConverter involved

### ISSUE THREAD

**Issue by kikaragyozov** [dotnet/runtime#102984]
If a consumer has defined a custom JsonConverter for an interface, to allow polymorphism, for example something like this:

```
public class CustomInterfaceConverter : JsonConverter<ICustomInterface>
{
	public override ICustomInterface? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
	{
		// Convert the C# property name to the expected JSON property name based on JsonSerializerOptions
		string typePropertyName = options.PropertyNamingPolicy?.ConvertName("Type") ?? "Type";

		using (JsonDocument doc = JsonDocument.ParseValue(ref reader))
		{
			JsonElement root = doc.RootElement;
			JsonElement typeProp;

			// Check for property name, considering case insensitivity
			if (options.PropertyNameCaseInsensitive)
			{
				if (!TryGetPropertyCaseInsensitive(root, typePropertyName, out typeProp))
				{
					throw new JsonException($"Type property '{typePropertyName}' is missing");
				}
			}
			else
			{
				if (!root.TryGetProperty(typePropertyName, out typeProp))
				{
					throw new JsonException($"Type property '{typePropertyName}' is missing");
				}
			}

			CustomInterfaceType type = (CustomInterfaceType)typeProp.GetInt32();
			return type switch
			{
				CustomInterfaceType.One => JsonSerializer.Deserialize<One>(root.GetRawText(), options),
				CustomInterfaceType.Two => JsonSerializer.Deserialize<Two>(root.GetRawText(), options),
				CustomInterfaceType.Three => JsonSerializer.Deserialize<Three>(root.GetRawText(), options),
				_ => throw new JsonException("Unknown type."),
			};
		}
	}

	public override void Write(Utf8JsonWriter writer, ICustomInterface value, JsonSerializerOptions options)
	{
		JsonSerializer.Serialize(writer, value, value.GetType(), options);
	}

	private static bool TryGetPropertyCaseInsensitive(JsonElement element, string name, out JsonElement value)
	{
		foreach (JsonProperty prop in element.EnumerateObject())
		{
			if (string.Equals(prop.Name, name, StringComparison.OrdinalIgnoreCase))
			{
				value = prop.Value;
				return true;
			}
		}

		value = default;
		return false;
	}
}
```

And you then try to return it as an `IAsyncEnumerable<ICustomInterface>`, you end up getting an `OutOfMemoryException` because the content is never flushed by the `IAsyncEnumerableOfTConverter<TAsyncEnumerable, TElement>`.

# Workarounds

Using `IAsyncEnumerable<object>` without a custom `JsonConverter<>` works as expected. You can then use the custom `JsonConverter<>` in the client code to consume it as an interface type, like so: 
```
await foreach (var message in JsonSerializer.DeserializeAsyncEnumerable<ICustomInterface>(stream, jsonOptions)
{
   // Your logic here.
}
```

Is this expected?

**Comment by eiriktsarpalis** [comment]
Presumably each individual element is too large to fit in-memory? This is a known restriction of custom converters that do not support streaming, and is not directly related to IAsyncEnumerable. This is being tracked by this issue https://github.com/dotnet/runtime/issues/63795

**Comment by kikaragyozov** [comment]
> Presumably each individual element is too large to fit in-memory? This is a known restriction of custom converters that do not support streaming, and is not directly related to IAsyncEnumerable. This is being tracked by this issue #63795

Each individual element is not too large to fit into memory. But rather, ShouldFlush never returns true and the collection keeps getting but into a buffer, and never being flushed out to the caller.

**Comment by BrennanConroy** [comment]
You mentioned `ShouldFlush` but also show using `DeserializeAsyncEnumerable` which are unrelated. `ShouldFlush` is only used for Serializing, not Deserializing.

Ignoring that discrepancy... from your code you have a similar pattern in your Write method to a similar issue https://github.com/dotnet/runtime/issues/66102 which has been fixed in 9.0.
